File: java/omnia/data/structure/Graph.java
package omnia.data.structure;

import java.util.Optional;

public interface Graph<E> {

  interface Node<E> {

    E item();

    Set<? extends Edge<E>> edges();

    Set<? extends Node<E>> neighbors();
  }

  interface Edge<E> {

    Collection<? extends Node<E>> endpoints();
  }

  Optional<? extends Node<E>> nodeOf(Object item);

  Set<E> contents();

  Set<? extends Node<E>> nodes();

  Set<? extends Edge<E>> edges();
}

File: src/omnia/data/structure/Graph.kt
package omnia.data.structure

interface Graph<E : Any> {

  interface Node<E : Any> {

    fun item(): E

    fun edges(): Set<out Edge<E>>

    fun neighbors(): Set<out Node<E>>
  }

  interface Edge<E : Any> {

    fun endpoints(): Collection<out Node<E>>
  }

  fun nodeOf(item: E): Node<E>?

  fun nodeOfUnknownType(item: Any?): Node<E>?

  fun contents(): Set<E>

  fun nodes(): Set<out Node<E>>

  fun edges(): Set<out Edge<E>>
}